### PR TITLE
refactor progress set to etcd/raft style

### PR DESCRIPTION
**Pull request by hicqu**
It's extracted from #317 , only refactor `ProgressSet` in this PR, so it should be more easily to review.

Signed-off-by: qupeng <qupeng@pingcap.com>

### CODE REVIEW

File: src/progress/progress_set.rs [hunk 1]
```diff
@@ -43,91 +44,98 @@ pub fn majority(total: usize) -> usize {
...
+    learners: Vec<u64>,
 
-    /// Create a new `Configuration` from a given `ConfState`.
-    pub fn from_conf_state(conf_state: &ConfState) -> Self {
-        Self {
-            voters: conf_state.voters.iter().cloned().collect(),
-            learners: conf_state.learners.iter().cloned().collect(),
-        }
+    // Demoted learners in joint consensus, because `learners` shouldn't intersect with `voters`.
+    learners_next: Vec<u64>,
```
Comment (Hoverbear) [inline src/progress/progress_set.rs]: In `voters` we use `[Vec<u64>; 2]`, can we not do this with `learners`?

File: src/progress/progress_set.rs [hunk 2]
```diff
@@ -149,114 +157,59 @@ pub enum CandidacyStatus {
...
+        let mut prs = HashMap::default();
+        for id in self.voters().chain(self.learners()) {
+            prs.insert(id, pr.clone());
         }
-
+        self.progress = prs;
         self.assert_progress_and_configuration_consistent();
     }
 
-    /// Returns the status of voters.
```
Comment (Hoverbear) [inline src/progress/progress_set.rs]: 👍

File: src/progress/progress_set.rs (hunk 1 above)
Comment (hicqu) [reply src/progress/progress_set.rs]: Of course we can. However, etcd/raft uses `learners_next`, so I guess it's better to keep it.

File: src/progress/progress_set.rs [hunk 3]
```diff
@@ -43,91 +44,98 @@ pub fn majority(total: usize) -> usize {
...
-                "There must be at least one voter.".into(),
-            ))
-        } else {
-            Ok(())
+    /// Create a new `Configuration` from a given `ConfState`.
+    pub fn from_conf_state(conf_state: &ConfState) -> Self {
+        Self::from(conf_state.clone())
+    }
+
+    // Test the configuration is valid or not. It's invalid when
```
Comment (Hoverbear) [inline src/progress/progress_set.rs]: I suggest you use `///` for better IDE support

**Comment by Fullstop000** [comment]
What's the progress of this? It looks like this has been paused for a long time

**Comment by BusyJay** [comment]
Joint consensus is fully implemented.

**Pull request closed without merging**